modules/axtask/src/wait_queue.rs

```diff
@@ -29,42 +40,46 @@ use crate::{current_run_queue, select_run_queue, AxTaskRef, CurrentTask};
 /// assert_eq!(VALUE.load(Ordering::Relaxed), 1);
 /// ```
 pub struct WaitQueue {
-    queue: SpinNoIrq<VecDeque<AxTaskRef>>,
+    list: SpinNoIrq<List<Node>>,
 }
 
-pub(crate) type WaitQueueGuard<'a> = SpinNoIrqGuard<'a, VecDeque<AxTaskRef>>;
+pub(crate) type WaitQueueGuard<'a> = SpinNoIrqGuard<'a, List<Node>>;
 
 impl WaitQueue {
     /// Creates an empty wait queue.
     pub const fn new() -> Self {
         Self {
-            queue: SpinNoIrq::new(VecDeque::new()),
+            list: SpinNoIrq::new(List::new()),
         }
     }
 
-    /// Creates an empty wait queue with space for at least `capacity` elements.
-    pub fn with_capacity(capacity: usize) -> Self {
-        Self {
-            queue: SpinNoIrq::new(VecDeque::with_capacity(capacity)),
-        }
+    fn waiter_blocked(&self, waiter: &Node) {
+        let curr = waiter.inner();
+        assert!(curr.is_running());
+        assert!(!curr.is_idle());
+        let mut wq_locked = self.list.lock();
+        debug!("task block: {}", curr.id_name());
+        // Mark the task as blocked, this has to be done before adding it to the wait queue
+        // while holding the lock of the wait queue.
+        curr.set_state(TaskState::Blocked);
+        wq_locked.push_back(waiter.clone());
     }
 
-    /// Cancel events by removing the task from the wait queue.
+    /// Cancel events by removing the task from the wait list.
```

Comment:
It's fine to call it a **queue** as the term “wait queue” is a proper noun. And the task is also FIFO when call wait and notify_one.

Reply:
Recover